logprep/processor/decoder/decoders.py

```diff
@@ -0,0 +1,219 @@
+"""implemented decoders"""
+
+import base64
+import binascii
+import re
+from functools import partial
+
+import msgspec
+
+from logprep.util.helper import FieldValue
+
+json_decoder = msgspec.json.Decoder()
+
+
+class DecoderError(Exception):
+    """raised if decoding fails"""
+
+
+def _parse(log_line, regex):
+    """this parses a given log_line to a dict via provided regex
+    it is really important that you don't use lookahead or lookbehinds
+    in the provided regex, because they could break the application by
+    recursive expressions in strings
+    """
+
+    def _parse(log_line, regex):
+        result = re.match(regex, log_line)
+        if result is None:
+            raise DecoderError("regex does not match")
+        return result.groupdict()
+
+    if isinstance(regex, re.Pattern):
+        return _parse(log_line, regex)
+
+    result = None
+    for r in regex:
+        try:
+            result = _parse(log_line, r)
+            break
+        except DecoderError:
+            pass
```

Comment:
I am aware that this kind of pattern is rather common in Python. Still, I would argue that (happy path) control flow using exceptions should be avoided, when possible

Reply:
Using zero cost exception here is better than asking for permission from a performance perspective